paging/paging-common/src/test/kotlin/androidx/paging/PagingDataDifferTest.kt

```diff
@@ -163,6 +163,339 @@ class PagingDataDifferTest(
+        differ.refresh()
+        assertThat(uiReceivers[0].refreshEvents).hasSize(1)
+        assertThat(uiReceivers[1].refreshEvents).hasSize(0)
+
+        // to recreate a real use-case of retry based on load error
+        pagingSources[1].errorNextLoad = true
+        loadDispatcher.executeAll()
+        // differ should still have first gen presenter
+        assertThat(differ.snapshot()).containsExactlyElementsIn(0 until 12)
+
+        // retry should be sent to gen 2 even though it wasn't presented
+        differ.retry()
+        assertThat(uiReceivers[0].retryEvents).hasSize(0)
+        assertThat(uiReceivers[1].retryEvents).hasSize(1)
+        loadDispatcher.executeAll()
+
+        // will retry with the correct cached hint
+        assertThat(differ.snapshot()).containsExactlyElementsIn(8 until 17)
+
+        // gen 2 receiver was recipient of the initial hint
+        assertThat(hintReceivers[1].hints).containsExactlyElementsIn(
+            listOf(
+                ViewportHint.Initial(
+                    presentedItemsBefore = 4,
+                    presentedItemsAfter = 4,
+                    originalPageOffsetFirst = 0,
+                    originalPageOffsetLast = 0,
+                )
+            )
+        )
+    }
+
+    @Test
+    fun refreshAfterStaticList() = testScope.runTest {
+        val differ = SimpleDiffer(dummyDifferCallback)
+
+        val pagingData1 = PagingData.from(listOf(1, 2, 3))
+        val job1 = launch { differ.collectFrom(pagingData1) }
+        assertTrue(job1.isCompleted)
+        assertThat(differ.snapshot()).containsAtLeastElementsIn(listOf(1, 2, 3))
+
+        val uiReceiver = UiReceiverFake()
+        val pagingData2 = infinitelySuspendingPagingData(uiReceiver = uiReceiver)
+        val job2 = launch { differ.collectFrom(pagingData2) }
+        assertTrue(job2.isActive)
+
+        // even though the second paging data never presented, it should be receiver of the refresh
+        differ.refresh()
+        assertThat(pagingData1.uiReceiver).isEqualTo(PagingData.NOOP_UI_RECEIVER)
+        assertThat(uiReceiver.refreshEvents).hasSize(1)
+
+        job2.cancel()
+    }
+
+    @Test
+    fun retryAfterStaticList() = testScope.runTest {
+        val differ = SimpleDiffer(dummyDifferCallback)
+
+        val pagingData1 = PagingData.from(listOf(1, 2, 3))
+        val job1 = launch { differ.collectFrom(pagingData1) }
+        assertTrue(job1.isCompleted)
+        assertThat(differ.snapshot()).containsAtLeastElementsIn(listOf(1, 2, 3))
+
+        val uiReceiver = UiReceiverFake()
+        val pagingData2 = infinitelySuspendingPagingData(uiReceiver = uiReceiver)
+        val job2 = launch { differ.collectFrom(pagingData2) }
+        assertTrue(job2.isActive)
+
+        // even though the second paging data never presented, it should be receiver of the retry
+        differ.retry()
+        assertThat(pagingData1.uiReceiver).isEqualTo(PagingData.NOOP_UI_RECEIVER)
```

Comment:
Ditto here for removing this assert that is always going to be true.

Reply:
done